Add configuration for encryption providers

sakshamsharma opened the pull request:
## Additions

Allows providing a configuration file (using flag `--experimental-encryption-provider-config`) to use the existing AEAD transformer (with multiple keys) by composing mutable transformer, prefix transformer (for parsing providerId), another prefix transformer (for parsing keyId), and AES-GCM transformers (one for each key). Multiple providers can be configured using the configuration file.

Example configuration:
```
kind: EncryptionConfig
apiVersion: v1
resources:
  - resources:
    - namespaces
    providers:
    - aes:
        keys:
        - name: key1
          secret: c2vjcmv0iglzihnly3vyzq==
        - name: key2
          secret: dghpcybpcybwyxnzd29yza==
    - identity: {}
```

Need for configuration discussed in:
#41939
[Encryption](https://github.com/destijl/community/blob/3418b4e4c6358f5dc747a37b90a97bc792f159ee/contributors/design-proposals/encryption.md)

**Pathway of a read/write request**:
1. MutableTransformer
2. PrefixTransformer reads the provider-id, and passes the request further if that matches.
3. PrefixTransformer reads the key-id, and passes the request further if that matches.
4. GCMTransformer tries decrypting and authenticating the cipher text in case of reads. Similarly for writes.

## Caveats
1. To keep the command line parameter parsing independent of the individual transformer's configuration, we need to convert the configuration to an `interface{}` and manually parse it in the transformer. Suggestions on better ways to do this are welcome.

2. Flags `--encryption-provider` and `--encrypt-resource` (both mentioned in [this document](https://github.com/destijl/community/blob/3418b4e4c6358f5dc747a37b90a97bc792f159ee/contributors/design-proposals/encryption.md) ) are not supported in this because they do not allow more than one provider, and the current format for the configuration file possibly supersedes their functionality.

3. Currently, it can be tested by adding `--experimental-encryption-provider-config=config.yml` to `hack/local-up-cluster.sh` on line 511, and placing the above configuration in `config.yml` in the root project directory.

Previous discussion on these changes:
https://github.com/sakshamsharma/kubernetes/pull/1

@jcbsmpsn @destijl @smarterclayton

## TODO
1. Investigate if we need to store keys on disk (per [encryption.md](https://github.com/destijl/community/blob/3418b4e4c6358f5dc747a37b90a97bc792f159ee/contributors/design-proposals/encryption.md#option-1-simple-list-of-keys-on-disk))
2. Look at [alpha flag conventions](https://github.com/kubernetes/kubernetes/blob/master/pkg/features/kube_features.go)
3. Need to reserve `k8s:enc` prefix formally for encrypted data. Else find a better way to detect transformed data.

Commit:
Add configuration options for encryption providers

Add location transformer, config for transformers

Location transformer helps choose the most specific transformer for
read/write operations depending on the path of resource being accessed.

Configuration allows use of --experimental-encryption-provider-config
to set up encryption providers. Only AEAD is supported at the moment.

Add new files to BUILD, AEAD => k8s-aes-gcm

Use group resources to select encryption provider

Update tests for configuration parsing

Remove location transformer

Allow specifying providers per resource group in configuration

Add IdentityTransformer configuration option

Fix minor issues with initial AEAD implementation

Unified parsing of all configurations

Parse configuration using a union struct

Run configuration parsing in APIserver, refactor parsing

More gdoc, fix minor bugs

Add test coverage for combined transformers

Use table driven tests for encryptionconfig

jcbsmpsn commented:
@k8s-bot ok to test

jcbsmpsn commented:
/assign smarterclayton

jcbsmpsn commented on staging/src/k8s.io/apiserver/pkg/server/options/encryption_provider_config.go:
```diff
@@ -0,0 +1,104 @@
+		if err != nil {
+			return err
+		}
+		if providerConfig.Kind == "AEAD" {
```
s/AEAD/k8s-aes-gcm/

sakshamsharma replied:
Is there a reason for skipping the `-v1` suffix here?

jcbsmpsn commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -48,6 +50,63 @@ func NewGCMTransformer(block cipher.Block) value.Transformer {
+		}), nil
+
+	} else {
+		return nil, fmt.Errorf("no valid keys found in configuration for AEAD transformer")
```
s/AEAD/k8s-aes-gcm-v1/

jcbsmpsn commented:
/release-note-none

jcbsmpsn commented:
/assign @smarterclayton

destijl commented on staging/src/k8s.io/apiserver/pkg/storage/value/location_transformer.go:
```diff
@@ -0,0 +1,97 @@
+				maxMatch = length
+				ts = []int{i}
+			} else if length == maxMatch {
+				ts = append(ts, i)
```
is there any point in using a slice? Even if there are two the same you just arbitrarily pick the first one. Seems like you could remove this line and it would work exactly the same.

sakshamsharma replied:
So what you suggest will be correct if you need to do writes. But if you're doing a read, you may need to try out all the transformers which could have written to that location in the past. So for instance, someone configures a transformer1 to write to location A. Then, some days later, the person wants to migrate to transformer2 for location A. There would still be some resources which were written with transformer1, and have not been migrated over yet. Thus we need to return all the transformers which *could* have written to that location, to be able to decrypt the information.

Now there is an issue with my implementation as well, something I had not noticed earlier:
If you initially use transformer1 for writing to location `/a/`, and then add transformer2 for writing to location `/a/b`, when you try to read `/a/b/c`, this function would not return transformer1 in the list. So we shall not be able to decrypt the resource because the TransformFromStorage shall now not use transformer1. I had not noticed this issue earlier.

An approach may be to try all (there won't be many, in most cases) transformers while reading, and only one transformer (what you suggested) while writing. Might have a performance impact, possibly.

Another approach would be to design a good migration scheme which shall detect a change in configuration, and will auto-migrate data to the new configuration. That way we know exactly which transformer was used to write to each location. I suppose this will require persisting the last configuration on disk in plaintext probably, and checking the current configuration against that at cluster bootup (we could, of course, omit the key secrets from this backup).

destijl commented on staging/src/k8s.io/apiserver/pkg/storage/value/location_transformer.go:
```diff
@@ -0,0 +1,97 @@
+}
+
+// TransformFromStorage collects all possible transformers (based on their location parameter) and
+// attempts to decrypt using each of them
```
guessing seems dangerous here. Do we need to add the resource type into the ciphertext spec with the key id?

sakshamsharma replied:
I've put my thoughts on this in the comment on `longestMatchingTransformers`

destijl commented on staging/src/k8s.io/apiserver/pkg/server/options/encryption_provider_config_test.go:
```diff
@@ -0,0 +1,136 @@
+func TestEncryptionProviderConfigCorrectParsing(t *testing.T) {
+	testConfigFile := ".test-enc-provider-correct"
+
+	err := ioutil.WriteFile(testConfigFile, []byte(correctConfig), 0600)
```
You could put the parsing code in another function that accepts io.Reader and pass buffers here, avoiding needing to actually write these to disk. Then the top level function just opens the file and passes it to the function that does the actual parsing.

sakshamsharma replied:
Will do that.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryption_provider_config_test.go:
```diff
@@ -0,0 +1,136 @@
+      secret: dGhpcyBpcyBwYXNzd29yZA==
+    - name: key3
+      secret: azhzIHNlY3JldCBzdG9yZQ==
+  resource: /registry/
```
We shouldn't be exposing etcd keys in this config.  This config should be able kubernetes resource types.  `pkg/storage/storage_factory.go` should accept a group resource override that includes a transformer - that transformer should be global.

I don't see a strong reason to do anything more complicated than identifying specific resources to encrypt (by opaquely passing in an encryption provider).  We deliberately do not expose details of storage location because they are mutable.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/aes/aes.go:
```diff
@@ -48,6 +50,63 @@ func NewGCMTransformer(block cipher.Block) value.Transformer {
+func NewGCMTransformerFromConfig(config map[string]interface{}) (value.Transformer, error) {
+
+	// Obtain list of keys as []interface{}
+	if keysInterface, ok := config["keys"].([]interface{}); ok {
```
We generally don't do this - we define API types and include the config that way.  Anything that reads from an API config should be outside this package (consuming it) and result in a properly instantiated prefix transformer.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/value/helpers.go:
```diff
@@ -0,0 +1,46 @@
+}
+
+// GetKeyDataFromConfig parses a valid interface to a KeyConfig object
+func GetKeyDataFromConfig(config interface{}) (KeyConfig, error) {
```
This is too generic - it's better to have a package that takes a concrete config api (i.e. Kube API structs) type and then knows how to turn that into a concrete AES PrefixTransformer, and then have  a fairly top level package in apiserver be consumable (pkg/storage/encryption/options.go) to pass into the StorageConfig.

You can define your config api types in a place  where config apis would generally belong - however, I know we are trying not to add more things to kubernetes/pkg/apis/componentconfig.  @deads2k where would api server specific config types go today?

sakshamsharma replied:
I suppose my previous comment (alternative configuration style) can help address this. Let me know if it is not so.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/value/location_transformer.go:
```diff
@@ -0,0 +1,97 @@
+)
+
+// LocationTransformer holds a transformer interface and the location of the resources it has to be used for.
+type LocationTransformer struct {
```
As mentioned above I don't think we should be doing any sort of transformation by key.  A storage is configured with a value transformer - we have 1 storage per api resource, and the level of encryption should be at one or more api resources, not at key level.

wojtek-t commented:
I had some email discussion about this with @sakshamsharma and I agree with @smarterclayton comments. I hopefully clarified a bunch of his questions offline.

sakshamsharma commented:
I have updated the code to allow specifying the resource names instead of the path now, as suggested by @smarterclayton . Thanks @wojtek-t !
Let me know if that portion looks fine now. I've also acted on @destijl 's recommendations.

As for the configuration comments, I am waiting on the request to @deads2k in the review above. I agree that the current way of parsing the configuration in `storage/value/encrypt/aes` is sub-optimal, but some more explanation on where to add the config-parsing would be a great help.

smarterclayton commented:
The key goal is to keep the transformer from knowing about config.  The mapping of config -> valid transformer setup is at a layer above, and then the api server at the top level needs to cleanly map from it's inputs to the config of the transformer.  We specifically encourage dependencies that are implementation specific (the AES impl vs a hypothetical secret box impl) to happen at the very highest level - where api server decides to use it based on its input switches.

Also note that config is an api - the code that defines apiserver config is at minimum in pkg/apiserver, not in pkg/apimachinery.  So the layering of those projects is important to preserve.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryption_provider_config.go:
```diff
@@ -0,0 +1,130 @@
+		if err != nil {
+			return err
+		}
+		if providerConfig.Kind == "k8s-aes-gcm" {
```
i'd say `k8s-aead-aes-gcm`, or we pick a colloquial name like `k8s-aes` and potentially version it in the future (for instance, if we add secret box, it might be `k8s-chacha`, and if we move to AES-GCM-SIV, it might be `k8s-aes-2`).  I think it's important to be specific about the choices, but most people don't understand the tradeoffs well enough that the additional info helps.

sakshamsharma replied:
The name was taken from [encryption.md](https://github.com/destijl/community/blob/3418b4e4c6358f5dc747a37b90a97bc792f159ee/contributors/design-proposals/encryption.md#aes-gcm-encryption-provider). So should I do `k8s-aes` ?

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryption_provider_config_test.go:
```diff
@@ -0,0 +1,102 @@
+// Throw error if invalid key for AES
+func TestEncryptionProviderConfigInvalidKey(t *testing.T) {
+	var destination map[schema.GroupResource]value.Transformer
+	if ConfigToTransformerOverrides(strings.NewReader(incorrectConfigInvalidKey), &destination) == nil {
```
I would avoid the map[string]string indirection here - the thing that instantiates your transformer should take a typed struct that matches the values from disk.  Basically, don't try to avoid the standard use of API machinery to construct config.

smarterclayton replied:
Note that for the first round of this I'd probably be ok with you just having a struct that matches your JSON schema and using json.Unmarshal.  Then validate and pass on.  We can easily refactor that into a formal API to move out of alpha, and it reduces the complexity of what you're trying to do before DCUT

sakshamsharma replied:
The reason the parsing had been done manually was that we need to accommodate different types of configurations. For instance, we know that the AES transformer needs a list of keys. The KMS transformer shall need an api-token. Hashicorp vault may need some different information. We needed to parse the YAML on the basis of the `kind:` key's value.

Instead, we came up with an alternative configuration:
```
- k8s-aes-v1:
    keys:
      - name: key1
        secret: c2VjcmV0IGlzIHNlY3VyZQ==
  resource: namespaces
- k8s-kms-v1:
    apitoken: foobar
  resource: secrets
```

We can parse a struct:
```
type Config struct {
    aesConfig AesConfig `yaml:"k8s-aes-v1"`
    kmsConfig KmsConfig `yaml:"k8s-kms-v1"`
    vaultConfig VaultConfig `yaml:"k8s-vault-v1"`
    resource string `yaml:"resource"`
}
```
And later check which one of them was not nil. The idea is to be able to parse the configuration differently on the basis of a particular property.

Another benefit of this one is that we can now keep AesConfig, KmsConfig and other such structs along with the code for the transformer, which hides the configuration details of the transformers from the upper layers. The versioning can be easily handled by the yaml key name as well.

I tried a prototype of the above parsing. Seems feasible.

smarterclayton replied:
If we're going to create new API objects, I'd prefer they resemble our existing API structures.  What I meant by alternative was following the convention, and having a simple list of which ones are supported.

I don't think the goal is to hide configuration structure or which implementations are available from upper levels like cmd/kube-apiserver.  Rather, goal is to hide config details from the lower levels like prefix transformer and storage.

I'd recommend staying with `kind` and `apiVersion`, supporting only two types right now (identity and aes-gcm) and then in 1.8 we can add real API objects for each of those configs, at which point we can reuse all of our infrastructure.

smarterclayton replied:
i think it’s ok to have an API object in the future with multiple types within it - the normal way we’d handle this config is

```
kind: EncryptionConfig
apiVersion: …
- resources: secrets,<other>
  providers:
  - identity: {}
  - aesgcm:
    keys:
    - name: key1
      secret: ...
    - name: key2
      secret: ...
```

smarterclayton replied:
In that model you'd use:

```
type Provider struct {
    AESGCM *AesConfig `json:"aesgcm"`
    KMS *KmsConfig `json:"kms"`
    Vault *VaultConfig `json:"vault"`
    Resources []string `json:"resources"`
}
```

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryption_provider_config_test.go:
```diff
@@ -0,0 +1,102 @@
+- kind: k8s-aes-gcm
+  version: v1
+  keys:
+    - name: key1
```
As I noted on the proposal, we need a way to switch from encrypted to not encrypted, which means the "first" key needs to indicate to the user that no encryption is to be performed (i.e. you rotate out of being encrypted the same way you rotate new keys in).  I don't know what actual mechanism should be used.

sakshamsharma replied:
If we keep an `unencrypted-key`, we may have to add a special case to not add the `<provider-name>:` prefix to the data while doing TransformToStorage for cases when data is unencrypted. Rather, it may be preferable to use a separate provider to achieve that.

Something like this (if I've understood this correctly):
```
- kind: identity
  version: v1
  resource: namespaces,secrets
- kind: k8s-aes-gcm
  version: v1
  keys:
    - name: key1
      secret: c2VjcmV0IGlzIHNlY3VyZQ==
    - name: key2
      secret: dGhpcyBpcyBwYXNzd29yZA==
  resource: namespaces,secrets,serviceaccounts
```

This way, identity transformer would be used for namespaces and secrets while encrypting to disk. But while reading from disk, we can still use k8s-aes, which seems to be the desired option. We can later add an endpoint to re-store all the data with the updated transformers, which would do a complete migration. That would be useful for rotation as well.

smarterclayton replied:
technically the identity transformer controls the prefix `{` and `k8s[\x00-\xFF]` already.  We need to prevent via validation people specifying that for any encryption provider except the default.

smarterclayton replied:
So we always need to support rotation across encryption schemes, as well as between no-scheme.  So it makes sense to order the top level types by this.  However, the scheme should be specified per resource, since you could in theory want to rotate per type.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryption_provider_config_test.go:
```diff
@@ -0,0 +1,102 @@
+      secret: c2VjcmV0IGlzIHNlY3VyZQ==
+    - name: key2
+      secret: dGhpcyBpcyBwYXNzd29yZA==
+  resource: /registry/namespaces
```
What's this here for?

sakshamsharma replied:
Sorry, shall fix in next commit.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/config/aes_config.go:
```diff
@@ -0,0 +1,73 @@
+}
+
+// Config contains the API configuration for an AES transformer
+type AesConfig struct {
```
style: we always capitalize leading acronyms/initialisms (`AESConfig`) in go code.  When it's private, you'd do `aesConfig`

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/config/config.go:
```diff
@@ -0,0 +1,97 @@
+/*
```
change the package to `options/encryptionconfig/config.go`

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/config/config.go:
```diff
@@ -0,0 +1,97 @@
+			Transformer: value.IdentityTransformer,
+			Prefix:      []byte{},
+		}
+	} else if parsedName.Type == "aes" {
```
this abstraction is unnecessary - just use:

```
type Provider struct {
  AES *AESConfig `json:"aes,omitempty"`
  Identity *IdentityConfig `json:"identity,omitempty"`
}
```

and in validation verify that only one of them is set.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go:
```diff
@@ -49,6 +49,10 @@ type identityTransformer struct{}
 
 func (identityTransformer) TransformFromStorage(b []byte, ctx Context) ([]byte, bool, error) {
+	// Other providers do not have k8s prefix
+	if len(b) > 0 && b[0] != '{' && bytes.HasPrefix(b, []byte("k8s:enc:")) {
```
This is no longer really the identity transformer (truly generic), it's one that is specific to encryption because it is checking against the encryption prefix.  It would be better to have this in a separate package dedicated to that, and add a better comment about the known types.  It's somewhat of an encapsulation violation because this method now has to know about all possible encodings, but I can live with that for a while.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -0,0 +1,116 @@
+
+// GetPrefixTransformer returns the PrefixTransformer from the provider configuration,
+// and returns an error if more than one providers were specified
+func (config *ProviderConfig) GetPrefixTransformer() (value.PrefixTransformer, error) {
```
We don't put methods on API objects.  These should be separate from the config.  Everything that is a struct that represents config should be in a single file types.go (see pkg/apis/componentconfig for an example), and anything that reads config should take those as inputs.  You should generate a Transformer from the top level config, but this is all stuff that has nothing to do with the objects.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -0,0 +1,116 @@
+type TransformerConfig interface {
+	// GetPrefixTransformer constructs and returns a PrefixTransformer using the provided configuration
+	GetPrefixTransformer() (value.PrefixTransformer, error)
+	// IsNil abstracts the nil check for the interface, since nil pointer checks on interface don't work
```
This shouldn't be necessary.  You really just want to iterate over a concrete config and then return a prefix transformer - I doubt this interface is necessary.

sakshamsharma replied:
We may add more transformers later. Checking everything with an if..else might get cumbersome. This interface lets us do things like [this](https://github.com/kubernetes/kubernetes/pull/46460/files/fa0d6f9187ec23cfb9c93f05d4b622dd9005d892#diff-1a93878ff6822de0984f38970a2298abR47) which may be helpful in the long run.

```
	for _, provider := range []TransformerConfig{config.AES, config.Identity} {
		if !provider.IsNil() {
                          provider.GetPrefixtransformer()                       
```

This way, the configuration is easily extensible I believe. I might be interpreting your comment incorrectly though.

sakshamsharma replied:
Is this what you mean here?
```
	if config.AES != nil {
		result, err = GetAESPrefixTransformer(config.AES)
		found = true
		if err != nil {
			return result, err
		}
	}

	if config.Identity != nil {
		if found == true {
			return result, fmt.Errorf("more than one provider specified in a single element, should split into different list elements")
		}
		result, err = GetIdentityPrefixTransformer()
		found = true
		if err != nil {
			return result, err
		}
	}

	if found == false {
		return result, fmt.Errorf("invalid provider configuration provided")
	}
```

smarterclayton replied:
effectively yes, and double check for both being set.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryption_provider_config.go:
```diff
@@ -0,0 +1,102 @@
+	}
+	defer f.Close()
+
+	if err := ConfigToTransformerOverrides(f, e.TransformerOverrides); err != nil {
```
We generally don't parse when flags are set.  We parse afterwards in a concrete step in the initialization code.  This method is probably just a function invoked out of this package from somewhere in the generic api server startup code.

smarterclayton reviewed (changes requested):
inline

sakshamsharma commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -495,6 +495,14 @@ func BuildStorageFactory(s *options.ServerRunOptions) (*serverstorage.DefaultSto
 		storageFactory.SetEtcdLocation(groupResource, servers)
 	}
 
+	transformerOverrides, err := s.Etcd.EncryptionProviderConfig.GetTransformerOverrides()
```
@smarterclayton Now the parsing is being done in the startup of the server. Is this what was recommended?

sakshamsharma commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+// GetTransformerOverrides returns the transformer overrides by reading and parsing the encryption provider configuration file
+func (config Config) GetTransformerOverrides() (map[schema.GroupResource]value.Transformer, error) {
```
I am not sure if this needs to be a method or a function. Should not be too hard to make this a function instead.

I made it a method because `s.Etcd.EncryptionProviderConfig.GetTransformerOverrides()` looked clean.

smarterclayton replied:
Generally we always take functions for things like this, we never add methods.  Your config is a path to a file, so:

```
func ParseTransformerConfig(filepath string) (map[schema.GroupResource]valueTransformer, error) {
```

would be our preferred method.

sakshamsharma commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+// ParseEncryptionConfiguration parses configuration data and returns the transformer overrides
+func ParseEncryptionConfiguration(f io.Reader) (map[schema.GroupResource]value.Transformer, error) {
```
Separated the actual parsing from file reading so we don't have to write the configuration to file during tests. Was suggested by @destijl

smarterclayton replied:
Yes this is fine

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -0,0 +1,179 @@
+}
+
+// GetPrefixTransformer constructs and returns the appropriate prefix transformers for the passed resource using its configuration
+func GetPrefixTransformers(config ResourceConfig) ([]value.PrefixTransformer, error) {
```
Take a pointer to ResourceConfig here.  Only pass by value if performance / use case warrants it.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -0,0 +1,179 @@
+	// Create a PrefixTransformer which shall later be put in a list with other providers
+	result = value.PrefixTransformer{
+		Transformer: keyTransformer,
+		Prefix:      []byte("k8s:enc:aes:v1:"),
```
Extract this as a private constant in this file

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go:
```diff
@@ -0,0 +1,50 @@
+limitations under the License.
+*/
+
+package encryptionconfig
```
Looks good, add more godoc describing what each field means.

sakshamsharma replied:
Resolved.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go:
```diff
@@ -0,0 +1,50 @@
+package encryptionconfig
+
+// File stores the complete configuration for encryption providers
+type File struct {
```
Call this top level type EncryptionConfig.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/identity/identity.go:
```diff
@@ -0,0 +1,50 @@
+	// EncryptIdentityTransformer has to return an error if the data is encoded using another transformer.
+	// JSON data starts with '{'. Protobuf data has a prefix 'k8s[\x00-\xFF]'.
+	// Prefix 'k8s:enc:' is reserved for encrypted data on disk.
+	if bytes.HasPrefix(b, []byte("k8s:enc:")) {
```
How will someone configure the encryption config to stop encrypting?  If identity transformer is first, won't this error out?  Did you add a fallback decoder to the prefix transformer codepath?

sakshamsharma replied:
Oh right. I had it handled in the location transformer, but it is not there at the moment.

Current prefix transformer code:
```
	for i, transformer := range t.transformers {
		if bytes.HasPrefix(data, transformer.Prefix) {
			result, stale, err := transformer.Transformer.TransformFromStorage(data[len(transformer.Prefix):], context)
			return result, stale || i != 0, err
		}
	}
	return nil, false, t.err
```

Proposed:
```
result, stale, err := transformer.Transformer.TransformFromStorage(data[len(transformer.Prefix):], context)
if len(transformer.Prefix) != 0 || err == nil {
	return result, stale || i != 0, err
}
```

Any transformer which has no prefix shall match all data. Thus, we need to ensure that such a transformer does not prevent other transformers from decrypting the data. If that transformer can decrypt without an error, it should be fine.

sakshamsharma replied:
Resolved.

sakshamsharma commented:
The tests seem to have been behaving weirdly, they are working now.

Need comments on this:

```
result, stale, err := transformer.Transformer.TransformFromStorage(data[len(transformer.Prefix):], context)
if len(transformer.Prefix) != 0 || err == nil {
	return result, stale || i != 0, err
}
```

Any transformer which has no prefix shall match all data. Thus, we need to ensure that such a transformer does not prevent other transformers from decrypting the data. If that transformer can decrypt without an error, it should be fine.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/storage/value/transformer.go:
```diff
@@ -126,7 +126,9 @@ func (t *prefixTransformers) TransformFromStorage(data []byte, context Context)
 		if bytes.HasPrefix(data, transformer.Prefix) {
 			result, stale, err := transformer.Transformer.TransformFromStorage(data[len(transformer.Prefix):], context)
-			return result, stale || i != 0, err
+			if len(transformer.Prefix) != 0 || err == nil {
```
can you invert this so it is a continue.  That's easier to read.  Generally exceptional cases are what we indent.  Also, add a comment about what the intent is.

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/encryptionconfig_test.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+// Throw error if invalid key for AES
+func TestEncryptionProviderConfigInvalidKey(t *testing.T) {
```
More tests that the returned transformer does what you expect.  Include the test for switching from encrypted to identity and vice versa.

sakshamsharma replied:
Resolved.

sakshamsharma reviewed (commented):
It appears the failing tests are flaky. In any case, this PR should not break any existing tests because there are no changes in the critical path of other tests (except one really minor one).

Do we need to re-run the tests?

@smarterclayton

fejta commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this
ref https://github.com/kubernetes/test-infra/pull/2931

smarterclayton commented:
@k8s-bot pull-kubernetes-unit test this

smarterclayton commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/encryptionconfig_test.go:
```diff
@@ -0,0 +1,193 @@
+		t.Fatalf("error while transforming data to storage using AES transformer: %s", err)
+	}
+
+	aesUntransformedData, stale, err = aesFirstTransformer.TransformFromStorage(identityTransformedData, context)
```
This cries out for a table driven test - might be better to restructure in that form.  Or at least shorter error messages :)

sakshamsharma commented:
Rebased from latest master. Converted tests to table driven (somewhat).

Not sure what the approval process is, but if it is possible to get it finished today, it would be great because the decision on the extension would take a week more.

smarterclayton commented:
Go ahead and squash

/approve

smarterclayton commented:
Lgtm pending squash

sakshamsharma commented:
@smarterclayton Done now. Sorry for the delay, was away from keyboard.

smarterclayton commented:
/lgtm

Although the queue is fairly stopped up

sakshamsharma commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

sakshamsharma commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

smarterclayton commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

sakshamsharma commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

sakshamsharma commented:
/retest

sakshamsharma commented:
Looking at the history (here)[https://submit-queue.k8s.io/#/history].

Apparently some commit broke the e2e test. All the recent PRs seem to have failed e2e tests.

enj commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:
```diff
@@ -0,0 +1,178 @@
+)
+
+const (
+	aesTransformerPrefixV1 = "k8s:enc:aes:v1:"
```
So we use this prefix to determine if and how we encrypted an object on disk?  Wouldn't it be easier to use something like `go-jose` / `JWE` to store this information in a tamper proof way?  We may still need `k8s:enc:` to prevent any collisions, but the algorithm would not need to be part of the prefix.

smarterclayton commented:
No it would not be easier :)  It be a long term approach, but we still have
to prefix and version.  JWT is much more complex at our existing storage
layer today with backwards compatibility taken into account.

On Mon, Jun 5, 2017 at 8:06 AM, Mo Khan <notifications@github.com> wrote:

> *@enj* commented on this pull request.
> ------------------------------
>
> In staging/src/k8s.io/apiserver/pkg/server/options/
> encryptionconfig/config.go
> <https://github.com/kubernetes/kubernetes/pull/46460#discussion_r120094954>
> :
>
> > +	"encoding/base64"
> +	"fmt"
> +	"io"
> +	"io/ioutil"
> +	"os"
> +
> +	yaml "github.com/ghodss/yaml"
> +
> +	"k8s.io/apimachinery/pkg/runtime/schema"
> +	"k8s.io/apiserver/pkg/storage/value"
> +	aestransformer "k8s.io/apiserver/pkg/storage/value/encrypt/aes"
> +	"k8s.io/apiserver/pkg/storage/value/encrypt/identity"
> +)
> +
> +const (
> +	aesTransformerPrefixV1 = "k8s:enc:aes:v1:"
>
> So we use this prefix to determine if and how we encrypted an object on
> disk? Wouldn't it be easier to use something like go-jose / JWE to store
> this information in a tamper proof way? We may still need k8s:enc: to
> prevent any collisions, but the algorithm would not need to be part of the
> prefix.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/46460#pullrequestreview-42029216>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p85Iblb5DmWSWaGIYdXKFojsyiAoks5sA-8xgaJpZM4Nm2qw>
> .
>

smarterclayton commented:
/retest

On Mon, Jun 5, 2017 at 9:25 AM, Clayton Coleman <ccoleman@redhat.com> wrote:

> No it would not be easier :)  It be a long term approach, but we still
> have to prefix and version.  JWT is much more complex at our existing
> storage layer today with backwards compatibility taken into account.
>
> On Mon, Jun 5, 2017 at 8:06 AM, Mo Khan <notifications@github.com> wrote:
>
>> *@enj* commented on this pull request.
>> ------------------------------
>>
>> In staging/src/k8s.io/apiserver/pkg/server/options/encryptionco
>> nfig/config.go
>> <https://github.com/kubernetes/kubernetes/pull/46460#discussion_r120094954>
>> :
>>
>> > +	"encoding/base64"
>> +	"fmt"
>> +	"io"
>> +	"io/ioutil"
>> +	"os"
>> +
>> +	yaml "github.com/ghodss/yaml"
>> +
>> +	"k8s.io/apimachinery/pkg/runtime/schema"
>> +	"k8s.io/apiserver/pkg/storage/value"
>> +	aestransformer "k8s.io/apiserver/pkg/storage/value/encrypt/aes"
>> +	"k8s.io/apiserver/pkg/storage/value/encrypt/identity"
>> +)
>> +
>> +const (
>> +	aesTransformerPrefixV1 = "k8s:enc:aes:v1:"
>>
>> So we use this prefix to determine if and how we encrypted an object on
>> disk? Wouldn't it be easier to use something like go-jose / JWE to store
>> this information in a tamper proof way? We may still need k8s:enc: to
>> prevent any collisions, but the algorithm would not need to be part of the
>> prefix.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/46460#pullrequestreview-42029216>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/ABG_p85Iblb5DmWSWaGIYdXKFojsyiAoks5sA-8xgaJpZM4Nm2qw>
>> .
>>
>
>

smarterclayton commented:
We'll need to update this based on
https://github.com/kubernetes/kubernetes/pull/46916

On Mon, Jun 5, 2017 at 7:44 PM, Kubernetes Submit Queue <
notifications@github.com> wrote:

> Merged #46460 <https://github.com/kubernetes/kubernetes/pull/46460>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/46460#event-1110879592>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABG_p3oqEyJSOoGgIDB11t27dCfiqfDbks5sBJLFgaJpZM4Nm2qw>
> .
>

php-coder commented on staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go:
```diff
@@ -388,7 +388,7 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
 	elems := make([]*elemForDecode, 0, len(getResp.Kvs))
 	for _, kv := range getResp.Kvs {
-		data, _, err := s.transformer.TransformFromStorage(kv.Value, authenticatedDataString(key))
+		data, _, err := s.transformer.TransformFromStorage(kv.Value, authenticatedDataString(kv.Key))
```
@sakshamsharma Could you explain what this change does, please?

sakshamsharma replied:
The intention is to have the location of the resource as the key. Since the key was written using its own location, it must be authenticated (while decrypting) using its own location as well. `kv.Key` is its location. `key` was a top level location (example: `/api/Secrets/`). When I was attempting to write integration tests (which couldn't get done at that time though), I discovered this issue.

Let me know if I went wrong somewhere.

php-coder commented on staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/identity/identity.go:
```diff
@@ -0,0 +1,50 @@
+
+// encryptIdentityTransformer performs no transformation on provided data, but validates
+// that the data is not encrypted data during TransformFromStorage
+type identityTransformer struct{}
```
@sakshamsharma @smarterclayton This is the 3rd `identityTransformer` in our codebase.

1) Do we really need to have 3 of them? What is the reason?

 2) I see that this implementation has check for `k8s:enc:` prefix. Why 2 others don't have such check? Should we copy it to them?

sakshamsharma replied:
This new identity transformer needs to return an error if it finds something that is encrypted. Without that check, the PrefixTransformer cannot try to use other transformers, because it will never see this one failing.

I had originally just modified the one in transformer.go to do this, but @smarterclayton reasoned that the identity transformer (which is used for all the cases where there is no encryption) should only be returning the data. Which is why, we now have a separate one for cases where we want to add more logic.

php-coder commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/encryptionconfig_test.go:
```diff
@@ -0,0 +1,172 @@
+/*
```
@sakshamsharma @smarterclayton I've just noticed that test named `encryptionconfig_test.go` while source file has name `config.go`. I can prepare a patch and rename it, if needed.

php-coder replied:
The file will be renamed in https://github.com/kubernetes/kubernetes/pull/53249

sakshamsharma commented on staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/encryptionconfig_test.go:
```diff
@@ -0,0 +1,172 @@
+/*
```
The package name is `encryptionconfig`. It may be violating some convention, but I'm not entirely sure.
